core/repos/repomanager.go

```diff
@@ -0,0 +1,389 @@
+				cleanErr := cleanCloneParentDirs(repo.getCloneParentDirs())
+				if cleanErr != nil {
+					return errors.New(err.Error() + " Failed to clean directories: " + cleanErr.Error())
+				}
+				return err
+			}
+		}
+
+		manager.repoList[repo.GetIdentifier()] = repo
+
+		// Set default repo
+		if len(manager.repoList) == 1 {
+			manager.setDefaultRepo(repo)
+		}
+	} else {
+		return errors.New("Repo already present")
+	}
+
+	return nil
+}
+
+func cleanCloneParentDirs(parentDirs []string) error {
+	for _, dir := range parentDirs {
+		if empty, err := isEmpty(dir); empty {
+			if err != nil {
+				return err
+			}
+
+			err = os.Remove(dir)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func isEmpty(path string) (bool, error) { //TODO: Teo, should this be put somewhere like utils?
+	dir, err := os.Open(path)
+	if err != nil {
+		return false, err
+	}
+	defer dir.Close() //Read-only we don't care about the return value
+
+	_, err = dir.Readdirnames(1)
+	if err == io.EOF {
+		return true, nil
+	}
+
+	return false, err
+}
+
+func (manager *RepoManager) GetRepos() (repoList map[string]*Repo) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	return manager.repoList
+}
+
+func (manager *RepoManager) RemoveRepoByIndex(index int) (ok bool, newDefaultRepo string) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	keys := manager.GetOrderedRepolistKeys()
+
+	for i, repoName := range keys {
+		if i != index {
+			continue
+		}
+		wasDefault := manager.repoList[repoName].Default
+		delete(manager.repoList, repoName)
+		// Set as default the repo sitting on top of the list
+		if wasDefault && len(manager.repoList) > 0 {
+			manager.setDefaultRepo(manager.repoList[manager.GetOrderedRepolistKeys()[0]]) //Keys have to be reparsed since there was a removal
+			newDefaultRepo = keys[0]
+		}
+		return true, newDefaultRepo
+	}
+
+	return false, newDefaultRepo
+}
+
+func (manager *RepoManager) RemoveRepo(repoPath string) (ok bool) { //Unused
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	if !strings.HasSuffix(repoPath, "/") { //Add trailing '/'
+		repoPath += "/"
+	}
+
+	repo, exists := manager.repoList[repoPath]
+	if exists {
+		wasDefault := repo.Default
+		delete(manager.repoList, repoPath)
+		// Set as default the repo sitting on top of the list
+		if wasDefault && len(manager.repoList) > 0 {
+			manager.setDefaultRepo(manager.repoList[manager.GetOrderedRepolistKeys()[0]])
+		}
+		return true
+	} else {
+		return false
+	}
+}
+
+func (manager *RepoManager) RefreshRepos() error {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	for _, repo := range manager.repoList {
+
+		err := repo.refresh()
+		if err != nil {
+			return errors.New("refresh repo for " + repo.GetIdentifier() + ":" + err.Error())
+		}
+	}
+
+	return nil
+}
+
+func (manager *RepoManager) RefreshRepo(repoPath string) error {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+
+	if !strings.HasSuffix(repoPath, "/") { //Add trailing '/'
```

Comment:
You seem to do this often. Fine as long as it works, but I'd suggest a tiny utility function for ensuring the trailing slash is there. Compilers these days are good enough to inline such situations.